sdk/python/lib/pulumi/x/automation/stack.py

```diff
@@ -0,0 +1,442 @@
+        """
+        Tries to create a new stack using the given workspace and stack name if the stack does not already exist,
+        or falls back to selecting the existing stack. If the stack does not exist,
+        it will be created and selected.
+
+        :param stack_name: The name identifying the Stack
+        :param workspace: The Workspace the Stack was created from.
+        :return: Stack
+        """
+        return Stack(stack_name, workspace, StackInitMode.CREATE_OR_SELECT)
+
+    def __init__(self, name: str, workspace: Workspace, mode: StackInitMode) -> None:
+        self.name = name
+        self.workspace = workspace
+
+        if mode is StackInitMode.CREATE:
+            workspace.create_stack(name)
+        elif mode is StackInitMode.SELECT:
+            workspace.select_stack(name)
+        elif mode is StackInitMode.CREATE_OR_SELECT:
+            try:
+                workspace.create_stack(name)
+            except StackAlreadyExistsError:
+                workspace.select_stack(name)
+
+    def up(self,
+           parallel: Optional[int] = None,
+           message: Optional[str] = None,
+           target: Optional[List[str]] = None,
+           expect_no_changes: Optional[bool] = None,
+           target_dependents: Optional[bool] = None,
+           replace: Optional[List[str]] = None,
+           on_output: Optional[OnOutput] = None,
+           program: Optional[PulumiFn] = None) -> UpResult:
+        """
+        Creates or updates the resources in a stack by executing the program in the Workspace.
+        https://www.pulumi.com/docs/reference/cli/pulumi_up/
+        """
+        program = program or self.workspace.program
+        extra_args = _parse_extra_args(**locals())
+        args = ["up", "--yes", "--skip-preview"]
+        args.extend(extra_args)
+
+        self.workspace.select_stack(self.name)
+        kind = ExecKind.LOCAL.value
+        on_exit = None
+
+        if program:
+            kind = ExecKind.INLINE.value
+            server = grpc.server(futures.ThreadPoolExecutor(max_workers=4),
+                                 options=_GRPC_CHANNEL_OPTIONS)
+            language_server = LanguageServer(program)
+            language_pb2_grpc.add_LanguageRuntimeServicer_to_server(language_server, server)
+
+            port = server.add_insecure_port(address="0.0.0.0:0")
+            server.start()
+
+            def on_exit():
+                language_server.on_pulumi_exit()
+                server.stop(0)
+            args.append(f"--client=127.0.0.1:{port}")
+
+        args.extend(["--exec-kind", kind])
+
+        try:
+            up_result = self._run_pulumi_cmd_sync(args, on_output)
+            outputs = self.outputs()
+            summary = self.info()
+            assert (summary is not None)
+            return UpResult(stdout=up_result.stdout, stderr=up_result.stderr, summary=summary, outputs=outputs)
+        except CommandError as exn:
+            if on_exit is not None:
```

Comment:
I think the cleanup routine in on_exit is something we want to make sure to call in both success and failure scenarios. Since we're dealing with funky global state, always resetting when we're done can help avoid hard to diagnose bugs in the future. 

For example, if the update finishes and the user for some reason decided to call `register_resource` outside of the inline program. In the success case, I believe there would still be a monitor configured. The user would be able to make the request, but the engine and monitor servers would be shut down and they would get an obscure rcp dial error.